File: ds_3_hash/L219.cpp
//
// Created by 李恒 on 2019/8/1.
//

//leetcode 219
//解决方法:滑动窗口+查找表

#include <iostream>
#include <unordered_set>
#include <vector>

using  namespace std;

class Solution{
public:
    bool containsNearbyDuplicate(vector<int>& nums , int k ){
        unordered_set<int> record;
        for(int i = 0 ; i < nums.size(); i++){
            if (record.find(nums[i]) != record.end())
                return true;
            record.insert(nums[i]);

            //始终保持查找表record中只有k个元素
            if(record.size() == k + 1)
                record.erase(nums[i - k]);  //删除record中最早插入的元素
        }
        return  false;
    }
};

File: python-algorithm/ds_3_hash/L219.py
# -*- coding: utf-8 -*-


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        d = {}
        for i in range(len(nums)):
            if nums[i] in d:
                if i-d[nums[i]] <= k:
                    return True
            # 每个值保存最大的索引即可，无序记录全部索引
            d[nums[i]] = i
        return False
